File: tests/integration/satisfies-operator-test.ts
// Satisfies Operator Test
// Tests TypeScript's satisfies operator for type validation without type narrowing

// Basic satisfies operator usage
const colors = {
  red: "#ff0000",
  green: "#00ff00",
  blue: "#0000ff",
} satisfies Record<string, string>;

// The type is still the specific object type, not Record<string, string>
// This allows us to access specific properties
const _redColor: string = colors.red;

// With satisfies, we can ensure an object matches a type while preserving its literal type
const config = {
  host: "localhost",
  port: 8080,
  ssl: false,
} satisfies {
  host: string;
  port: number;
  ssl: boolean;
};

// Type is preserved - we can access properties with their literal types
const _port: 8080 = config.port; // Type is 8080, not number

// Satisfies with const assertion
const routes = {
  home: "/",
  about: "/about",
  contact: "/contact",
} as const satisfies Record<string, string>;

// Function with satisfies
function getConfig() {
  return {
    apiUrl: "https://api.example.com",
    timeout: 5000,
    retryAttempts: 3,
  } satisfies {
    apiUrl: string;
    timeout: number;
    retryAttempts: number;
  };
}

// Variable declaration with satisfies
const apiConfig = getConfig();

// Satisfies with union types
type Status = "loading" | "success" | "error";
const currentStatus = "success" satisfies Status;

// Satisfies with arrays
const _numbers = [1, 2, 3, 4, 5] satisfies number[];
const _mixedArray = [1, "two", true, { x: 10 }] satisfies unknown[];

// Satisfies with nested objects
const user = {
  id: 1,
  name: "John Doe",
  settings: {
    theme: "dark",
    notifications: true,
  },
} satisfies {
  id: number;
  name: string;
  settings: {
    theme: string;
    notifications: boolean;
  };
};

// Satisfies with optional properties
interface UserProfile {
  username: string;
  email?: string;
  age?: number;
}

const profile = {
  username: "johndoe",
  email: "john@example.com",
} satisfies UserProfile;

// Satisfies with generic types
interface Container<T> {
  value: T;
  setValue: (val: T) => void;
}

const _stringContainer = {
  value: "hello",
  setValue: (val: string) => {
    console.log("Setting value:", val);
  },
} satisfies Container<string>;

// Complex satisfies with multiple constraints
type Theme = "light" | "dark" | "auto";
type Language = "en" | "es" | "fr";

const preferences = {
  theme: "dark",
  language: "en",
  fontSize: 14,
} satisfies {
  theme: Theme;
  language: Language;
  fontSize: number;
};

// Testing the results
console.log("Colors red:", colors.red);
console.log("Config port:", config.port);
console.log("Routes home:", routes.home);
console.log("API config:", apiConfig.apiUrl);
console.log("Current status:", currentStatus);
console.log("User name:", user.name);
console.log("Profile username:", profile.username);
console.log("Preferences theme:", preferences.theme);

console.log("Satisfies operator test completed");

File: tests/integration/satisfies-operator-test/satisfies-operator-test.cpp
#include "satisfies-operator-test.h"

using namespace js;

const js::any colors = []() {
      js::object obj_temp_0;
      obj_temp_0.set("red", "#ff0000"_S);
      obj_temp_0.set("green", "#00ff00"_S);
      obj_temp_0.set("blue", "#0000ff"_S);
      return js::any(obj_temp_0);
    }();

const js::string _redColor = colors["red"];

const js::any config = []() {
      js::object obj_temp_1;
      obj_temp_1.set("host", "localhost"_S);
      obj_temp_1.set("port", js::number(8080));
      obj_temp_1.set("ssl", false);
      return js::any(obj_temp_1);
    }();

const js::any _port = config["port"];

const js::any routes = []() {
      js::object obj_temp_2;
      obj_temp_2.set("home", "/"_S);
      obj_temp_2.set("about", "/about"_S);
      obj_temp_2.set("contact", "/contact"_S);
      return js::any(obj_temp_2);
    }();

auto getConfig() {
    return []() {
          js::object obj_temp_3;
          obj_temp_3.set("apiUrl", "https://api.example.com"_S);
          obj_temp_3.set("timeout", js::number(5000));
          obj_temp_3.set("retryAttempts", js::number(3));
          return js::any(obj_temp_3);
        }();
}

const js::any apiConfig = getConfig();

const js::string currentStatus = "success"_S;

js::array<js::number> _numbers = js::array<js::number>{js::number(1), js::number(2), js::number(3), js::number(4), js::number(5)};

js::array<js::number> _mixedArray = js::array<js::number>{js::number(1), "two"_S, true, []() {
      js::object obj_temp_4;
      obj_temp_4.set("x", js::number(10));
      return js::any(obj_temp_4);
    }()};

const js::any user = []() {
      js::object obj_temp_5;
      obj_temp_5.set("id", js::number(1));
      obj_temp_5.set("name", "John Doe"_S);
      obj_temp_5.set("settings", []() {
      js::object obj_temp_6;
      obj_temp_6.set("theme", "dark"_S);
      obj_temp_6.set("notifications", true);
      return js::any(obj_temp_6);
    }());
      return js::any(obj_temp_5);
    }();

const js::any profile = []() {
      js::object obj_temp_7;
      obj_temp_7.set("username", "johndoe"_S);
      obj_temp_7.set("email", "john@example.com"_S);
      return js::any(obj_temp_7);
    }();

const js::any _stringContainer = []() {
      js::object obj_temp_8;
      obj_temp_8.set("value", "hello"_S);
      obj_temp_8.set("setValue", [](js::string val) -> auto {
js::console.log("Setting value:"_S, val);
});
      return js::any(obj_temp_8);
    }();

const js::any preferences = []() {
      js::object obj_temp_9;
      obj_temp_9.set("theme", "dark"_S);
      obj_temp_9.set("language", "en"_S);
      obj_temp_9.set("fontSize", js::number(14));
      return js::any(obj_temp_9);
    }();

// Entry point
void Main() {
    js::console.log("Colors red:"_S, colors["red"]);
    js::console.log("Config port:"_S, config["port"]);
    js::console.log("Routes home:"_S, routes["home"]);
    js::console.log("API config:"_S, apiConfig->apiUrl);
    js::console.log("Current status:"_S, currentStatus);
    js::console.log("User name:"_S, user->name);
    js::console.log("Profile username:"_S, profile["username"]);
    js::console.log("Preferences theme:"_S, preferences["theme"]);
    js::console.log("Satisfies operator test completed"_S);
}

int main(int /*argc*/, char** /*argv*/) {
    Main();
    return 0;
}